tableaudocumentapi/datasource.py

```diff
@@ -59,6 +59,31 @@ def from_file(cls, filename):
         dsxml = ET.parse(filename).getroot()
         return cls(dsxml, filename)
 
+    def get_connections_by_server_name(self, server_name):
+        matches = []
+        if server_name:
+            for connection in self._connections:
+                if str(server_name) == connection.server:
+                    matches.append(connection)
+
+        return matches
+
+    def get_connections_by_attributes(self, conn_attributes):
+        from itertools import product
+
+        matches = []
+
+        valid_attributes = ('server', 'username', 'dbclass', 'dbname')
+
+        if not (all(attr in valid_attributes for attr in conn_attributes) and conn_attributes):
+            raise RuntimeError("That is not a valid connection attribute")
+
+        for attr, conn in product(conn_attributes, self._connections):
+            if getattr(conn, attr).lower() == conn_attributes[attr].lower():
+                matches.append(conn)
+
+        return matches
+
```

Comment:
This all feels unnecessary to me. Even with a federated datasource, the number of connections will be small. I would rather just give back a list of the connections and let people push that through a map or iterate themselves over these properties instead of creating special case methods that don't seem to add a ton of enhancements. What you have looks like a good sample but it feels early to start exposing such seemingly specialized functions. We can talk about why you think these methods merit being a part of the core classes we expose.